New function [CSVSink] - allowing to serialise Detections to a CSV file

AdonaiVera opened the pull request:
## Description

This PR introduces the `CSVSink` class to address structured logging of object detection data into CSV format, as requested in issue [#746](https://github.com/roboflow/supervision/issues/746). The class supports logging of bounding boxes, class IDs, confidence scores, and frame numbers, enhancing post-processing data analysis and interpretability.

The PR includes:
- `CSVSink` class implementation.
- Comprehensive unit tests for the class.
- A demo showcasing the integration of `CSVSink` with a video processing pipeline.

Note: The `CSVSink` class was placed directly in the utils folder in the file.py file due to its capability to handle various types of file operations integral to the project's core functionality. Open to suggestions if there's a more appropriate location within the project structure.

## Type of Change

- [x] New feature: Added `CSVSink` class for structured logging of detection data.

## Testing

The `CSVSink` class was validated through a series of unit tests, covering:
- Logging with various custom data scenarios.
- Handling of scenarios with no detections.
- Verification of CSV output format and data integrity.

A demo script demonstrates the class's usage within a video processing workflow, annotating video frames and logging detections to a CSV file.  [Demo Here](https://colab.research.google.com/drive/1DyKR_euGxHy4YXD_QGA6bwsBNDs9mScV?usp=sharing)

## Documentation Updates

- Added `CSVSink` class documentation within the module, detailing its purpose, usage, and example usage.
- Updated README to include a section on structured detection data logging, introducing the `CSVSink` class.

SkalskiP commented on supervision/utils/file.py:
```diff
@@ -1,10 +1,121 @@
+    ) -> None:
+        if not self.writer:
+            raise Exception(
+                f"Cannot append to CSV: The file '{self.filename}' is not open. Ensure that the 'open' method is called before appending data."
```
Let's break the line here. Make it under 88 characters. Otherwise, Ruff will complain.

SkalskiP commented on supervision/utils/file.py:
```diff
@@ -1,10 +1,121 @@
+        for i in range(len(detections.xyxy)):
+            self.write_detection_row(detections, i, custom_data)
+
+    def write_header(self, detections: Detections, custom_data: Dict[str, Any]) -> None:
```
Let's separate the parsing logic from the writing logic. The parsing method should be static. This will allow us to write unit tests for parsing separately. 

Let me know if this explanation makes sense. I can give you more details if something needs to be clarified.

AdonaiVera replied:
Hi @SkalskiP 👋 
Do you mean we should create a static method called `parse_detection_data` that takes `detections` and `custom_data` as inputs and returns a list of dictionaries, where each dictionary represents a row to be written to the CSV file? If it is like that, we can obtain these rows by calling the `parse_detection_data` function and write them to the CSV file using the `append` method.

Let me make a test quick and I'll update the PR 💪

SkalskiP commented on supervision/utils/file.py:
```diff
@@ -1,10 +1,121 @@
+        self.writer.writerow(self.fieldnames)
+        self.header_written = True
+
+    def write_detection_row(
```
Let's separate the parsing logic from the writing logic. The parsing method should be static. This will allow us to write unit tests for parsing separately. 

Let me know if this explanation makes sense. I can give you more details if something needs to be clarified.

SkalskiP commented:
Hi @AdonaiVera 👋🏻 ! Good start. I left a few comments. Let me know if you have any questions. I didn't look at unit tests yet. Let's work on logic first.

AdonaiVera commented on supervision/utils/file.py:
````diff
@@ -1,10 +1,121 @@
+
+    Usage:
+        ```python
+        from supervision.utils.detections import Detections
````
I actually developed the demo using a track object algorithm. [here](https://colab.research.google.com/drive/1DyKR_euGxHy4YXD_QGA6bwsBNDs9mScV?usp=sharing) 

Do you think an example like this could work? 
```python
        import numpy as np
        import supervision as sv
        from ultralytics import YOLO
        import time

        model = YOLO("yolov8n.pt")
        tracker = sv.ByteTrack()
        box_annotator = sv.BoundingBoxAnnotator()
        label_annotator = sv.LabelAnnotator()
        csv_sink = sv.CSVSink(...)

        def callback(frame: np.ndarray, _: int) -> np.ndarray:
            start_time = time.time()
            results = model(frame)[0]
            detections = sv.Detections.from_ultralytics(results)
            detections = tracker.update_with_detections(detections)

            labels = [
                f"#{tracker_id} {results.names[class_id]}"
                for class_id, tracker_id
                in zip(detections.class_id, detections.tracker_id)
            ]
            time_frame = (time.time() - start_time)

            csv_sink.append(detections, custom_data={"processing_time": time_frame})

            annotated_frame = box_annotator.annotate(
                frame.copy(), detections=detections)
            return label_annotator.annotate(
                annotated_frame, detections=detections, labels=labels)

        csv_sink.open()
        sv.process_video(
            source_path="people-walking.mp4",
            target_path="result.mp4",
            callback=callback
        )
        csv_sink.close()

``` 

I added an example of time spent processing a frame, but we can use any custom value.

SkalskiP replied:
I think the example in the docs " reference " section should be relatively minimal. We should not introduce elements like annotators in the CSV sink section of docs.

But we should create additional, more comprehensive tutorials showing how to combo those independent tools. But this is outside of the scope.

AdonaiVera commented:
Hi @SkalskiP 

Thank you for your valuable feedback. I have already implemented the changes and created a PR. Additionally, I updated the demo and moved the unit test to the test/detections folder. Please let me know your thoughts on the changes and if you have any further suggestions 🥷

SkalskiP commented:
As I already mentioned. For the method/function to be visible in docs. It needs to be public, and it needs to have docstring. `CSVSink` methods are undocumented, and as a result, users don't see them in docs. 

![Screenshot 2024-02-01 at 13 26 00](https://github.com/roboflow/supervision/assets/26109316/007996be-451a-4981-98d4-09e977a3d2c9)

SkalskiP commented on test/detection/test_csv.py:
```diff
@@ -0,0 +1,111 @@
+        sink.append(second_detections, second_custom_data)
+
+    # Read back the CSV file and verify its contents
+    with open(csv_filename, mode="r", newline="") as file:
```
Id create a function `assert_csv_equal` or something like this to confirm if csvs are equal.

SkalskiP commented on test/detection/test_csv.py:
```diff
@@ -0,0 +1,111 @@
+        [55, 65, 75, 85, 1, 0.9, 3, "car", 43],
+    ]
+
+    # Using the CSVSink class to write the detection data to a CSV file
```
You can remove those comments. :) Seems redundant.

AdonaiVera commented on mkdocs.yml:
```diff
@@ -41,6 +41,7 @@ nav:
         - Polygon Zone: detection/tools/polygon_zone.md
         - Inference Slicer: detection/tools/inference_slicer.md
         - Detection Smoother: detection/tools/smoother.md
+        - Save CSV Detection: detection/tools/csv_sink.md
```
I like the idea; we can combine the two functions 💪

SkalskiP replied:
To be precise. :) Not combine functions but use single docs page to document them. It is probably obvious but still l, want to make that clear. Haha

AdonaiVera replied:
yes, combine only the docs 💪 
`Save Detections: detection/tools/save_detections.md`

AdonaiVera commented on supervision/detection/tools/csv_sink.py:
```diff
@@ -0,0 +1,147 @@
+        label_annotator = sv.LabelAnnotator()
+        csv_sink = sv.CSVSink(...)
+
+        def callback(frame: np.ndarray, _: int) -> np.ndarray:
```
I followed your suggestion and simplified the example.
```python
import cv2
import supervision as sv
from ultralytics import YOLO

image = cv2.imread(<SOURCE_IMAGE_PATH>)
model = YOLO(...)

result = model(image)[0]
detections = sv.Detections.from_ultralytics(result)
with csv_sink as sink:
    sink.append(detections, custom_data={'frame_index': 1})
```
What do you think?

SkalskiP replied:
I updated your example with my feedback. I think a video example is better as (I think) people will primarily use `CSVSink` with video processing.

AdonaiVera replied:
Amazing, thank you @SkalskiP

AdonaiVera commented on supervision/detection/tools/csv_sink.py:
```diff
@@ -0,0 +1,147 @@
+            parsed_rows.append(row)
+        return parsed_rows
+
+    def append(
```
It's a really good point. What should we do if they change the `custom_data` or the `data` in the detections. Should we throw an exception in the append function? Or should we add a column to the CSV and then check all the empty values and fill them with `None`? What do you think? 

Thinking about the user experience, I think allowing changes and leaving the empty values could be useful. But let me know what you think.

SkalskiP replied:
We will not do something soft for now - inform users about inconsistency. We will replace it with `WARNING` once we implement this [task](https://github.com/roboflow/supervision/issues/835).

AdonaiVera commented on supervision/detection/tools/csv_sink.py:
```diff
@@ -0,0 +1,147 @@
+        self.close()
+
+    def open(self) -> None:
+        self.file = open(self.filename, "w", newline="")
```
Umm, I am going to add a function to create the folder if it doesn't already exist.

AdonaiVera commented on supervision/detection/tools/csv_sink.py:
```diff
@@ -0,0 +1,147 @@
+            self.file.close()
+
+    @staticmethod
+    def parse_detection_data(
```
Hi @SkalskiP, I added a condition that ensures the `class_id`, `confidence`, and `tracker_id` columns are always present in the CSV, even if they are missing for a particular detection (Optionals values), In case any of these columns are missing, the CSV file will still contain the respective column, but it will be empty. 

Please let me know if that is okay?

AdonaiVera replied:
During testing, I discovered that the `mock_detections` function does not support complex data structures similar to the one provided.
```python
data={"metadata": {"sensor_id": 101, "location": "north"}, "tags": ["urgent", "review"]}
``` 

I got the next issue
```python
test/detection/test_csv.py:112: in <module>
    mock_detections(
test/test_utils.py:19: in mock_detections
    return Detections(
<string>:9: in __init__
    ???
supervision/detection/core.py:61: in __post_init__
    validate_detections_fields(
supervision/detection/utils.py:591: in validate_detections_fields
    validate_data(data, n)
supervision/detection/utils.py:568: in validate_data
    raise ValueError(f"Shape of np.ndarray for key '{key}' must be ({n},)")
E   ValueError: Shape of np.ndarray for key 'tags' must be (1,)
```

AdonaiVera commented:
Hi @SkalskiP 👋  
I have made all the necessary changes. However, there are still some questions that I left on the thread. 

I tried to handle multiple scenarios, but I can work with more if needed. But before that, I need to solve the question about what happens if the custom data changes between detections.

Thanks 🚀

SkalskiP commented:
I made a few docs updates in the meantime. Nothing important. There was no point in forcing you to make corrections. These were very minor things.

SkalskiP commented:
Hi @AdonaiVera 👋🏻 I made two small commits to your branch today. I really wanted to close it out before the weekend. Feel free to ask questions here in comments.

Overal awesome work! Thanks a lot for your help! 🔥

AdonaiVera commented:
Thank you! We did an amazing collaborative work. @SkalskiP 💪  

Now that I have the full perspective, I'm ready to create the JSON saving function, I'll update that PR soon 🚀
